docs: make structural directives guide generic

dario-piotrowicz opened the pull request:
tweak the current structural directives guide (currently mainly targeted
at the creation of custom structural directives) so that is more generic
and a point of reference for structural directives in general

this also includes the re-addition of the one-per-element section
removed in PR #40015

resolves #44786

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #44786


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information

- @AndrewKushnir :slightly_smiling_face::+1: 

- Before & After:
  Before:
  ![Screenshot at 2022-01-29 21-24-14](https://user-images.githubusercontent.com/61631103/151678012-7ee4c401-adc7-4dcf-a9e7-b261703f2205.png)
  
  ![Screenshot at 2022-01-29 21-23-05](https://user-images.githubusercontent.com/61631103/151678016-3f11965f-292a-40ae-936c-abf7c0902326.png)

  After:
  ![Screenshot at 2022-01-29 21-24-45](https://user-images.githubusercontent.com/61631103/151678022-9a24ff83-7fb9-4b3a-8c42-45a7c185f3e7.png)
  
  ![Screenshot at 2022-01-29 21-22-24](https://user-images.githubusercontent.com/61631103/151678028-7db52622-4869-4dd5-a826-1049c0042e66.png)

- The "Improving template type checking for custom directives" section should be moved to a different guide, I figured I could keep things cleared by moving it away from the page in a follow-up PR 

- This partially undoes changes made in PR #40015

dario-piotrowicz committed:
docs: make structural directives guide generic

tweak the current structural directives guide (currently mainly targeted
at the creation of custom structural directives) so that is more generic
and a point of reference for structural directives in general

this also includes the re-addition of the one-per-element section
removed in PR #40015

resolves #44786

dario-piotrowicz commented on aio/content/examples/structural-directives/src/app/app.component.html:
```diff
@@ -110,15 +110,20 @@ <h2 id="ngFor">NgFor</h2>
 <p class="code">&lt;div *ngFor="let hero of heroes; let i=index; let odd=odd; trackBy: trackById" [class.odd]="odd"&gt;</p>
 <!--#docregion inside-ngfor -->
-<div *ngFor="let hero of heroes; let i=index; let odd=odd; trackBy: trackById" [class.odd]="odd">
+<div
```
I split this and the one below because in the UI the lines were easily wrapped, this should marginally help

dario-piotrowicz commented on aio/content/guide/built-in-directives.md:
```diff
@@ -240,15 +240,12 @@ then uses this template repeatedly to create a new set of elements and bindings
 in the list.
 For more information about shorthand, see the [Structural Directives](guide/structural-directives#shorthand) guide.
 
-{@a one-per-element}
```
this section isn't really explaining clearly the one-per-element topic, a section just for that has been added in the structural directives guide.

dario-piotrowicz commented on aio/content/guide/built-in-directives.md:
```diff
@@ -240,15 +240,12 @@ then uses this template repeatedly to create a new set of elements and bindings
-
-Because structural directives add and remove nodes from the DOM, apply only one structural directive per element.
 
-For more information about `NgFor` see the [NgForOf API reference](api/common/NgForOf).
```
this guide already talks about the `NgFor` in the previous sections, so putting here the link seems to me pretty unnecessary

dario-piotrowicz commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
 </div>
 
-For more information on Angular's built-in structural directives, such as `NgIf`, `NgForOf`, and `NgSwitch`, see [Built-in directives](guide/built-in-directives).
+## What they are
```
may not be the best section name ever.... :sweat:

bob-watson replied:
You could eliminate the need for this header by moving lines 13-15 to above the example. 
Because that's the obvious place to see a description of the subject you're about to discuss further in the topic, it doesn't need a heading to introduce it.

dario-piotrowicz commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
+</div>
+
+
+{@a one-per-element}
+## One structural directive per element
```
This section was removed in PR #40015 (which by the way is the PR that made this guide non-generic)

I am not a big fan of moving changes back and forth but I believe the way this was refactored really just lacks clarity and makes things much more difficult to understand.

I put back the section with one or two minor tweaks as I thought it was already well written, if we want I can rewrite this with my own words too.

dario-piotrowicz commented on aio/content/guide/structural-directives.md:
```diff
@@ -295,12 +309,14 @@ The value of the property can be either a general type-narrowing function based
 For example, consider the following structural directive that takes the result of a template expression as an input:
 
 <code-example language="ts" header="IfLoadedDirective">
-export type Loaded<T> = { type: 'loaded', data: T };
```
I updated this to use `&lt;` and `&gt;` since without them the type annotations are actually not displayed:
![Screenshot at 2022-01-29 21-34-55](https://user-images.githubusercontent.com/61631103/151678285-2c0d7a27-cd51-4e45-a1fa-937af888cae5.png)

(PS: I also split the lines to help in readability)

dario-piotrowicz commented on aio/content/guide/structural-directives.md:
```diff
@@ -87,73 +168,6 @@ To verify that the directive works, click the button to change the value of `con
-
-For more information, see the [NgFor API](api/common/NgForOf "API: NgFor") and [NgForOf API](api/common/NgForOf) documentation.
-
-### Creating template fragments with `<ng-template>`
-
-Angular's `<ng-template>` element defines a template that doesn't render anything by default.
-With `<ng-template>`, you can render the content manually for full control over how the content displays.
-
-If there is no structural directive and you wrap some elements in an `<ng-template>`, those elements disappear.
-In the following example, Angular does not render the middle "Hip!" in the phrase "Hip! Hip! Hooray!" because of the surrounding `<ng-template>`.
-
-<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (template-tag)" region="template-tag"></code-example>
-
-<div class="lightbox">
-  <img src='generated/images/guide/structural-directives/template-rendering.png' alt="template tag rendering">
-</div>
-
```
I don't think this should sit in the structural directives guide, I changed this (see line 70 above) to a quick alert redirecting to the ng-template api docs (in which I also added a shortened alert with the example shown here)

dario-piotrowicz commented:
I just noticed that I am applying the fix for: https://github.com/angular/angular/issues/40739

in that issue it is suggested to move the example to a doc region, my fixing it here shouldn't prevent that from being done afterwards anyways (maybe as a follow-up PR, as I mentioned in the PR's description the section should also be moved elsewhere, maybe both things can be done in the same PR)

AndrewKushnir commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
-{@a unless}
+Structural directives are directives which change the DOM layout by adding and removing DOM element.
+
+Angular offers a set of built-in structural directives which are widely used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).
```
```suggestion
Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).
```

AndrewKushnir commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
+
+<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (asterisk)" region="asterisk"></code-example>
+
+What Angular does in this case is creating an `<ng-template>` element and moving the `*ngIf` directive onto it where it becomes a property binding in square brackets, `[ngIf]`. The rest of the `<div>`, including its class attribute, is then moved inside the `<ng-template>`:
```
```suggestion
Angular creates an `<ng-template>` element and applies the `*ngIf` directive onto it where it becomes a property binding in square brackets, `[ngIf]`. The rest of the `<div>`, including its class attribute, is then moved inside the `<ng-template>`:
```

AndrewKushnir commented on aio/content/guide/structural-directives.md:
````diff
@@ -1,17 +1,98 @@
+Note that Angular does not actually create a real `<ng-template>` element, but instead only renders the `<div>` and adds a comment node placeholder to the DOM.
+
+```html
+<!--bindings={
+  "ng-reflect-ng-if": "[object Object]"
+}-->
+<div _ngcontent-c0>Mr. Nice</div>
+
````
I think we can omit HTML comments, they are not load bearing:
```suggestion
<div _ngcontent-c0>Mr. Nice</div>
```

dario-piotrowicz replied:
I think the comment was there just to show what's explained in the line above
> ... adds a comment node placeholder ...

so I actually don't thing that it is too bad, do you find this not a useful info for the user? should we also remove the text part talking about the comment node?

AndrewKushnir replied:
Yeah, I think we can remove the reference to the comment node. It's an implementation detail, so we should not highlight it in a guide.

AndrewKushnir commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
+
+<div class="alert is-helpful">
+
+  Note that Angular's `<ng-template>` element defines a template that doesn't render anything by default, if you simply wrap elements in an `<ng-template>` without applying a structural directive those elements will simply disappear.
```
```suggestion
  Note that Angular's `<ng-template>` element defines a template that doesn't render anything by default, if you just wrap elements in an `<ng-template>` without applying a structural directive those elements will not be rendered.
```

AndrewKushnir commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,98 @@
+{@a one-per-element}
+## One structural directive per element
 
+Someday you'll want to repeat a block of HTML but only when a particular condition is true.
+You'll _try_ to put both an `*ngFor` and an `*ngIf` on the same host element.
+Angular won't let you. You may apply only one _structural_ directive to an element.
+
+The reason is simplicity. Structural directives can do complex things with the host element and its descendants.
+When two directives lay claim to the same host element, which one should take precedence?
```
```suggestion
It's a quite common use-case to repeat a block of HTML but only when a particular condition is true. An intuitive way to do that is to put both an `*ngFor` and an `*ngIf` on the same element. However, since both `*ngFor` and an `*ngIf` are structural directives, this would be treated as an error by the compiler. You may apply only one _structural_ directive to an element.

The reason is simplicity. Structural directives can do complex things with the host element and its descendants.
When two directives lay claim to the same host element, which one should take precedence?
```

AndrewKushnir commented on aio/content/special-elements/core/ng-template.md:
```diff
@@ -2,6 +2,21 @@ Angular's `<ng-template>` element defines a template that is not rendered by def
 
+<div class="alter is-helpful">
+
+  Note that if you simply wrap content inside an `<ng-template>` without instructing Angular to render it, such content will simply not appear in your application. As you can see in the following example, Angular does not render the middle "Hip!" in the phrase "Hip! Hip! Hooray!" because of the surrounding `<ng-template>`.
```
```suggestion
  Note that if you wrap content inside an `<ng-template>` without instructing Angular to render it, such content will not appear on a page. As you can see in the following example, Angular does not render the middle "Hip!" in the phrase "Hip! Hip! Hooray!" because of the surrounding `<ng-template>`.
```

dario-piotrowicz replied:
I've updated the text, I've also noticed that "As you can see" part which actually made sense before when there was an image but doesn't anymore, so I restructured that sentence a bit :slightly_smiling_face:

AndrewKushnir reviewed (commented):
Awesome work @dario-piotrowicz 👍 

I've left a few comments, please let me know what you think.

Once the feedback is addressed, we'd need a docs team review of the content as well.

Thank you.

dario-piotrowicz committed:
fixup! docs: make structural directives guide generic

dario-piotrowicz commented:
@AndrewKushnir thanks a lot for your awesome reviewing! :heart_eyes: 

(and sorry for making you put that many comments :pray: :sweat:)

I've addressed your comments, please have another look :slightly_smiling_face:

atscott commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,95 @@
+# Structural directives
 
-This topic demonstrates how to create a structural directive and provides conceptual information on how directives work, how Angular interprets shorthand, and how to add template guard properties to catch template type errors.
+This guides talks about structural directives and provides conceptual information on how such directives work, how Angular interprets their shorthand syntax, and how to add template guard properties to catch template type errors.
```
```suggestion
This guide is about structural directives and provides conceptual information on how such directives work, how Angular interprets their shorthand syntax, and how to add template guard properties to catch template type errors.
```

dario-piotrowicz replied:
:facepalm::+1:

atscott commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,95 @@
-{@a unless}
+Structural directives are directives which change the DOM layout by adding and removing DOM element.
+
+Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects, for more information see [Built-in directives](guide/built-in-directives).
```
```suggestion
Angular provides a set of built-in structural directives (such as `NgIf`, `NgFor`, `NgSwitch` and others) which are commonly used in all Angular projects. For more information see [Built-in directives](guide/built-in-directives).
```

atscott commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,95 @@
+
+## Structural directive shorthand
+
+When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`, such convention is shorthand that Angular interprets and converts into a longer form.
```
```suggestion
When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`. This convention is shorthand that Angular interprets and converts into a longer form.
```

atscott commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,95 @@
+## Structural directive shorthand
+
+When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`, such convention is shorthand that Angular interprets and converts into a longer form.
+Substantially Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.
```
```suggestion
Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.
```

atscott commented on aio/content/guide/structural-directives.md:
```diff
@@ -1,17 +1,95 @@
+When structural directives are applied they generally are prefixed by an asterisk, `*`,  such as `*ngIf`, such convention is shorthand that Angular interprets and converts into a longer form.
+Substantially Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that surrounds the host element and its descendants.
+
+Let's for example take the following code which uses an `*ngIf` to displays the hero's name if `hero` exists:
```
```suggestion
For example, let's take the following code which uses an `*ngIf` to displays the hero's name if `hero` exists:
```

atscott commented:
Added @TeriGlover for editing review.

dario-piotrowicz committed:
fixup! docs: make structural directives guide generic

dario-piotrowicz commented:
Thanks for the nice reviewing @atscott :slightly_smiling_face:, all addressed :+1:

dario-piotrowicz commented:
Hi @TeriGlover, any change you could have a look at this PR? :bow: :pray: :bow:

AndrewKushnir commented:
@bob-watson could you please help with the review for this PR from docs perspective (Andrew and myself verified that the context is correct from the technical perspective)? Thank you.

bob-watson reviewed (commented):
If the "what they are" heading really bugs you, I offered an alternative.
Other than that, it seems reasonable after a quick glance.

dario-piotrowicz committed:
fixup! docs: make structural directives guide generic

AndrewKushnir commented:
Merge-assistance: the approvals got reset after a fixup commit. I've reviewed and approved it, so this PR is ready for merge.

alxhub commented:
This PR was merged into the repository by commit c8086b7e860d2b42fb83a2f3f92ee5a3e7a7c331.
